flow/display_list.cc

```diff
@@ -0,0 +1,1273 @@
+
+#undef DL_OP_DISPATCH
+
+      default:
+        FML_DCHECK(false);
+        return;
+    }
+  }
+}
+
+static bool CompareOps(uint8_t* ptrA,
+                       uint8_t* endA,
+                       uint8_t* ptrB,
+                       uint8_t* endB) {
+  if (endA - ptrA != endB - ptrB) {
+    return false;
+  }
+  uint8_t* bulkStartA = ptrA;
+  uint8_t* bulkStartB = ptrB;
+  while (ptrA < endA && ptrB < endB) {
+    auto opA = (const DLOp*)ptrA;
+    auto opB = (const DLOp*)ptrB;
+    if (opA->type != opB->type || opA->size != opB->size) {
+      return false;
+    }
+    ptrA += opA->size;
+    ptrB += opB->size;
+    FML_DCHECK(ptrA <= endA);
+    FML_DCHECK(ptrB <= endB);
+    dlCompare result;
+    switch (opA->type) {
+#define DL_OP_EQUALS(name)                              \
+  case DisplayListOpType::name:                         \
+    result = static_cast<const name##Op*>(opA)->equals( \
+        static_cast<const name##Op*>(opB));             \
+    break;
+
+      FOR_EACH_DISPLAY_LIST_OP(DL_OP_EQUALS)
+
+#undef DL_OP_DISPATCH
+
+      default:
+        FML_DCHECK(false);
+        return false;
+    }
+    switch (result) {
+      case dlCompare::kNotEqual:
+        return false;
+      case dlCompare::kUseBulkCompare:
+        break;
+      case dlCompare::kEqual:
+        // Check if we have a backlog of bytes to bulk compare and then
+        // reset the bulk compare pointers to the address following this op
+        auto bulkBytes = reinterpret_cast<const uint8_t*>(opA) - bulkStartA;
+        if (bulkBytes > 0) {
+          if (memcmp(bulkStartA, bulkStartB, bulkBytes) != 0) {
+            return false;
+          }
+        }
+        bulkStartA = ptrA;
+        bulkStartB = ptrB;
+        break;
+    }
+  }
+  if (ptrA != endA || ptrB != endB) {
+    return false;
+  }
+  if (bulkStartA < ptrA) {
+    // Perform a final bulk compare if we have remaining bytes waiting
+    if (memcmp(bulkStartA, bulkStartB, ptrA - bulkStartA) != 0) {
+      return false;
+    }
+  }
+  return true;
+}
+
+void DisplayList::renderTo(SkCanvas* canvas) const {
+  DisplayListCanvasDispatcher dispatcher(canvas);
+  dispatch(dispatcher);
+}
+
+bool DisplayList::equals(const DisplayList& other) const {
+  return CompareOps(ptr_, ptr_ + used_, other.ptr_, other.ptr_ + other.used_);
+}
+
+DisplayList::DisplayList(uint8_t* ptr,
+                         size_t used,
+                         int opCount,
+                         const SkRect& cull)
+    : ptr_(ptr),
+      used_(used),
+      opCount_(opCount),
+      bounds_({0, 0, -1, -1}),
+      boundsCull_(cull) {
+  static std::atomic<uint32_t> nextID{1};
+  do {
+    uniqueID_ = nextID.fetch_add(+1, std::memory_order_relaxed);
+  } while (uniqueID_ == 0);
+}
+
+DisplayList::~DisplayList() {
+  DisposeOps(ptr_, ptr_ + used_);
+}
+
+#define DL_BUILDER_PAGE 4096
+
+// copy_v(dst, src,n, src,n, ...) copies any number of typed srcs into dst.
+static void copy_v(void* dst) {}
```

Comment:
The routines here do not follow the style guide. Presumably, this is just from Skia's SkLiteDL implementation?

Reply:
Yes, cut and paste.

I also have ingested a few SkAssert calls that could probably change to FML_DCHECKs.

Reply:
I made quite a number of changes to style in https://github.com/flutter/engine/pull/26928/commits/cd7fed2fd769dccfc6735920b35ebd41db523cd0

You might want to review that commit in isolation and see if you agree with them. Many of the lines are dup/analogous changes so much of it can be scanned.

Reply:
I did a lot of style guide work over the past couple of days. I'm going to mark this conversation resolved and leave it for further review to mention specific issues.